### DISCUSSION TITLE
Switch to CMake?

### DISCUSSION THREAD

**Discussion by metalefty** [neutrinolabs/xrdp#1966] [Ideas]
Just starting a discussion.

The `configure.ac` script is a shell script. It grows as development progress. Actually, #1962 adds 51 lines to `configure.ac`.

Honestly, I like writing shell scripts but don't like maintaining shell scripts. Should we switch to CMake in long-term?

**Comment by hardening** [comment]
By `cmake` you mean `meson` I guess ;) ?

**Comment by matt335672** [comment]
I've not used cmake (or meson) for anything significant, so it's difficult for me to comment other than saying that any replacement will need to support very similar features (e.g.):-

- Available on our CI systems, including older Ubuntu (18.04) and FreeBSD
- Supports unit test framework with `make check` and `make distcheck`

I agree that `autotools` are fiddly, and may well deter developers used to something more modern.

@aquesnel  - I get a feeling from somewhere you may have something useful to add to this discussion.

**Reply by eli-schwartz** [reply]
both cmake and meson should be available on your CI systems... Ubuntu 18.04 has cmake 3.10.2 and meson 0.45.1 but:
- meson explicitly supports Bionic by capping the minimum version of python and requiring all CI to pass on Bionic, so you can just pip install the latest version
- cmake is a compiled binary and the official site has standalone versions which are very compatible

regarding dist:
- I think there are some things you can do with cpack but I'm no expert.
- I do know (quite well) that `meson dist` is designed to work out of the box and provide the same familiar usefulness that autotools does, e.g. it is available by default without configuration, it will generate a tarball for you and then unpack that and run a full configure/build/test/install cycle on that. There is no split between `make dist` and `make distcheck`, just `meson dist` and pass `--no-tests` if you don't want the "check" part

Disclaimer: I loved meson enough to join the core development team, so obviously I am inclined to shill that. Please tell me if you'd like me to try selling you on it. :D

**Comment by aquesnel** [comment]
I don't have a strong opinion on one build tool vs another. I think that changing the build system will be a lot of work, and that picking a build system can easily become a "solution in search of a problem". 

For me to be onboard with changing the build system I'd like to see a description of the problem the xrdp project is facing with `autotools`, as well as a list of requirements that `autotools` is fulfilling that we want to keep in the new build system. I think @matt335672 's list of requirements is a good place to start.

**Comment by hardening** [comment]
@eli-schwartz so fun to read you, as I had the same story as you: trying `meson`, loving it when using it on some projects, and finally pushing some improvements.
